File: 421_simplify-path/simplify-path.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/simplify-path
@Language: C++
@Datetime: 15-06-01 07:15
*/

class Solution {
public:
    /**
     * @param path the original path
     * @return the simplified path
     */
    string simplifyPath(string &path) {
		int i = 0, n = path.size();
		for (int j = 0; j < n; ) {
			if (path[j] != '.' || j + 1 < n && path[j + 1] != '/'  && path[j + 1] != '.'||
			 j + 2 < n && path[j + 1] != '/' && path[j + 2] != '/' && path[j + 2] != '.') {
				if (path[j] != '/' || i == 0 || path[i - 1] != '/' && j < n - 1)
					path[i++] = path[j++];
				else
					j++;
			}
			else {
				if (j + 1 < n && path[j + 1] == '.') {
					if (j + 2 < n && path[j + 2] == '.') { //handle ...
						for (int k = 0; k < 3; path[i++] = path[j++], k++);
					}
					else {//handle ..
						if (i > 1) { //only pop when current path is not root 
							for (--i; i > 1 && path[i - 1] != '/'; --i);
							if (i > 1) --i;
						}
						j += 2;
					}
				}
				else 
					j++;
			}
		}
		return path.substr(0, i);
	}	
};

File: 421_simplify-path/simplify-path.java
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/simplify-path
@Language: Java
@Datetime: 15-06-01 07:48
*/

public class Solution {
    /**
     * @param path the original path
     * @return the simplified path
     */
    public static String simplifyPath(String path) {
        // Write your code here
        String[] tokens = path.split("/+");
        if (tokens.length == 0) return "/";
        List<String> s = new ArrayList<String>();
        for (String t : tokens) {
        	if (t.equals("..")) {
        		if (s.size() > 1)
        			s.remove(s.size() - 1);
        	}
        	else if (!t.equals(".")) s.add(t);	
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.size(); ++i) {
        	sb.append(s.get(i));
        	if (i == 0 || i < s.size() - 1) sb.append("/");
        }
        return sb.toString();
    }
}
